decode.go

```diff
@@ -962,35 +969,59 @@ func unmarshalListValue(
 // Structs can represent either a message or a map.
 // For messages, the message descriptor can be provided or inferred from the node.
 // For maps, the field descriptor can be provided to validate the map keys/values.
-func unmarshalStruct(
-	unm *unmarshaler,
+func (u *unmarshaler) unmarshalStruct(
 	node *yaml.Node,
 	message *structpb.Struct,
 	msgDesc protoreflect.MessageDescriptor,
 	field protoreflect.FieldDescriptor,
+	ignoreDynamicType bool, // If true, ignore @type fields.
 ) {
 	if field != nil {
 		// Validate the field is a message or map.
 		if !field.IsMap() && field.Kind() != protoreflect.MessageKind {
-			unm.addErrorf(node, "expected %s for %v, got mapping", getExpectedNodeKind(field, false), field.FullName())
+			u.addErrorf(node, "expected %s for %v, got mapping", getExpectedNodeKind(field, false), field.FullName())
 		}
-	} else if msgDesc == nil {
+	} else if msgDesc == nil && !ignoreDynamicType {
 		// Try to find the message descriptor.
-		msgType, _ := unm.findAnyType(node)
+		msgType, _ := u.findAnyType(node)
 		if msgType != nil {
 			msgDesc = msgType.Descriptor()
 		}
 	}
 
+	// Check for a custom unmarshaler
+	if msgDesc != nil {
+		if custom, ok := u.custom[msgDesc.FullName()]; ok {
+			msgType, err := u.resolveType(msgDesc.FullName())
+			if err != nil {
+				u.addError(node, err)
+			} else if custom(u, node, msgType.New().Interface()) {
+				// Custom unmarshaler handled any errors, recursively.
+				msgDesc = nil
+				ignoreDynamicType = true
+			}
+		}
+	}
```

Comment:
So... if I follow this correctly: this package will enforce _additional_ structural checks on the contents of a `google.protobuf.Struct` if it contains a `"@type"` property that indicates a known type. Is that right? And this new code is just an optimization so that the case of an any inside an any is only verified once, at the root/top-level any, and not re-verified for nested ones?

This seems a bit surprising that this might refuse to parse a `google.protobuf.Struct` (which should be free-form JSON structure) because it thinks the data "looks like" an `Any` message and then does extra validation with the assumption that it **is** an `Any`.

Or am I totally misreading this?

Reply:
you are reading it correctly, and make some really good points. Thanks for reviewing it so carefully!